include/El/core/SyncInfo.hpp

```diff
@@ -0,0 +1,197 @@
+#ifndef EL_CORE_SYNCINFO_HPP_
+#define EL_CORE_SYNCINFO_HPP_
+
+#include "IndexSequence.hpp"
+
+namespace El
+{
+
+/** \class SyncInfo
+ *  \brief Manage device-specific synchronization information.
+ *
+ *  Device-specific synchronization information. For CPUs, this is
+ *  empty since all CPU operations are synchronous with respect to the
+ *  host. For GPUs, this will be a stream and an associated event.
+ *
+ *  The use-case for this is to cope with the matrix-free part of the
+ *  interface. Many of the copy routines have the paradigm that they
+ *  take Matrix<T,D>s as arguments and then the host will organize and
+ *  dispatch subkernels that operate on data buffers, i.e., T[]
+ *  data. In the GPU case, for example, this provides a lightweight
+ *  way to pass the CUDA stream through the T* interface without an
+ *  entire matrix (which, semantically, may not make sense).
+ *
+ *  This also might be useful for interacting with
+ *  Aluminum/MPI/NCCL/whatever. It essentially enables tagged
+ *  dispatch, where the tags possibly contain some extra
+ *  device-specific helpers.
+ */
+template <Device D> struct SyncInfo
+{
+    SyncInfo() {}
+
+    template <typename T>
+    SyncInfo(Matrix<T,D> const&) {}
+
+};// struct SyncInfo<D>
+
+template <Device D>
+void AddSynchronizationPoint(SyncInfo<D> syncInfo)
+{}
+
+template <Device D>
+void AddSynchronizationPoint(SyncInfo<D> A, SyncInfo<D> B)
+{}
+
+template <Device D>
+void Synchronize(SyncInfo<D> const&)
+{}
+
+#ifdef HYDROGEN_HAVE_CUDA
+
+template <>
+struct SyncInfo<Device::GPU>
+{
+    SyncInfo() : SyncInfo{GPUManager::Stream(), GPUManager::Event()} {}
+
+    SyncInfo(cudaStream_t stream, cudaEvent_t event)
+        : stream_{stream}, event_{event} {}
+
+    template <typename T>
+    SyncInfo(Matrix<T,Device::GPU> const& A)
+        : stream_{A.Stream()}, event_{A.Event()} {}
+
+    cudaStream_t stream_;
+    cudaEvent_t event_;
+};// struct SyncInfo<Device::GPU>
+
+inline void AddSynchronizationPoint(SyncInfo<Device::GPU> const& syncInfo)
+{
+    EL_CHECK_CUDA(cudaEventRecord(syncInfo.event_, syncInfo.stream_));
+}
+
+inline void AddSynchronizationPoint(
+    SyncInfo<Device::CPU> const& A, SyncInfo<Device::GPU> const& B)
+{
+    LogicError("I don't know what should happen here.");
+}
+
+inline void AddSynchronizationPoint(
+    SyncInfo<Device::GPU> const& A, SyncInfo<Device::CPU> const& B)
+{
+    LogicError("I don't know what should happen here.");
+}
+
+// This captures the work done on A and forces B to wait for completion
```

Comment:
I sorta hope nobody ever needs to look at the stream and event outside of this `SyncInfo` thing. But your concern is valid since I expose them in the `Matrix` class. Should I `AddSynchronizationPoint(A);` outside the `if` block?

Reply:
Semantically, I'm not sure that makes sense to guarantee from this function, since it says nothing about being able to later synchronize on it. If `A == B`, then this should be a NOP, and if you care about being able to synchronize `A` with something else later, you should use the single-argument version.

Reply:
I agree. Really all you're guaranteeing is "up-to-date"-ness and an object is implicitly always up to date with itself. IMO, one shouldn't synchronize to an event unless they know where it was last recorded; there is no guarantee about its state (hence why this function first records an event and then waits on it).